contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c

```diff
@@ -245,81 +289,98 @@ tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset)
-	 * fihish reading from/in the middle of the page (walreceiver) in such a
-	 * case we should preserve the last read page header for the IV data and
-	 * the encryption state.
-	 *
-	 * TODO: check if walsender/receiver restarts form the beggining of the
-	 * page in case of the crash.
-	 */
-	for (dec_off = 0; dec_off < readsz;)
+	if (!keys)
 	{
-		data_size = Min(page_size, readsz);
+		/* cache is empty, try to read keys from disk */
+		keys = pg_tde_fetch_wal_keys(0);
+	}
 
-		if (page_size == XLOG_BLCKSZ)
-		{
-			memcpy((char *) curr_page_hdr, (char *) buf + dec_off, SizeOfXLogShortPHD);
+#ifndef FRONTEND
+	write_key_lsn = pg_atomic_read_u64(&EncryptionState->enc_key_lsn);
+#endif
 
-			/* set the flag to "not encrypted" for the walreceiver */
-			((XLogPageHeader) ((char *) buf + dec_off))->xlp_info &= ~XLP_ENCRYPTED;
+	if (write_key_lsn != 0)
+	{
+		WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+		Assert(last_key);
 
-			Assert(curr_page_hdr->xlp_magic == XLOG_PAGE_MAGIC || curr_page_hdr->xlp_magic == 0);
-			dec_off += XLogPageHeaderSize(curr_page_hdr);
-			data_size -= XLogPageHeaderSize(curr_page_hdr);
-			/* it's a beginning of the page */
-			iv_ctr = 0;
-		}
-		else
+		/* write has generated a new key, need to fetch it */
+		if (last_key->start_lsn < write_key_lsn)
 		{
-			/* we're in the middle of the page */
-			iv_ctr = (offset % XLOG_BLCKSZ) - XLogPageHeaderSize(curr_page_hdr);
-		}
+			pg_tde_fetch_wal_keys(write_key_lsn);
 
-		if ((data_size + dec_off) > readsz)
-		{
-			data_size = readsz - dec_off;
+			/* in case cache was empty before */
+			keys = pg_tde_get_wal_cache_keys();
 		}
+	}
 
-		if (curr_page_hdr->xlp_info & XLP_ENCRYPTED)
+	SetXLogPageIVPrefix(tli, segno, iv_prefix);
+
+	XLogSegNoOffsetToRecPtr(segno, offset, wal_segsz_bytes, data_start);
+	XLogSegNoOffsetToRecPtr(segno, offset + count, wal_segsz_bytes, data_end);
+	
+	curr_key = keys;
+	while (curr_key)
+	{
+#ifdef TDE_XLOG_DEBUG
+		elog(DEBUG1, "WAL key %X/%X-%X/%X", LSN_FORMAT_ARGS(curr_key->start_lsn), LSN_FORMAT_ARGS(curr_key->end_lsn));
+#endif
+
+		if (curr_key->key->internal_key.start_lsn != InvalidXLogRecPtr &&
+				(curr_key->key->internal_key.rel_type & TDE_KEY_TYPE_WAL_ENCRYPTED))
 		{
-			if (key == NULL)
+			/* 
+			 * Check if the key's range overlaps with the buffer's and decypt
+			 * part that does.
+			 */
+			if (data_start <= curr_key->end_lsn && curr_key->start_lsn <= data_end)
 			{
-				key = GetTdeGlobaleRelationKey(GLOBAL_SPACE_RLOCATOR(XLOG_TDE_OID));
+				dec_off = XLogSegmentOffset(Max(data_start, curr_key->start_lsn), wal_segsz_bytes);
+				dec_sz = XLogSegmentOffset(Min(data_end, curr_key->end_lsn), wal_segsz_bytes) - dec_off;
+#ifdef TDE_XLOG_DEBUG
+				elog(DEBUG1, "decrypt WAL, dec_off: %lu [buff_off %lu], sz: %lu", dec_off, offset - dec_off, dec_sz);
+#endif
+				PG_TDE_DECRYPT_DATA(iv_prefix, dec_off,
+							(char *) buf + (offset - dec_off),
+							dec_sz, (char *) buf + (offset - dec_off),
+							curr_key->key);
```

Comment:
missing break here if we the current lsn of the key is > end lsn of data?

Reply:
yes